### ISSUE TITLE
What is the best way to get the predicted values of the training set in Darts?

### ISSUE THREAD

**Issue by ETTAN93** [unit8co/darts#1781]
I am trying to get the **predicted values of my training set** in Darts. In SKlearn, one can simply do:
```
model.fit(training_set)
model.predict(training_set)
```

What is the equivalent method in Darts assuming I have target lags, past covariate lags and future covariate lags?

From what I've tried, the .predict() method is only forward looking after you fit your data so I won't be able to get the predictions of my training set.

Thanks in advance.

**Comment by madtoinou** [comment]
Hi @ETTAN93,

If by "get the predicted values of my training set" you mean "generate a model forecast for the training set index (timestamps)", you'll need to use a little workaround since `model.predict()` uses the last timestamp of the training series to generate the index of the output:

```python
# get the predictable timestamps given the lags of the model
predictable_timestamps = model._get_historical_forecastable_time_index(series=train_set, past/future_covariates=..., is_training=False)
# have a dummy ts finish just before the first predictable timestamp
dummy_series, _ = training_set.split_before(predictable_timestamps[0])
# run inference ignoring the timeseries "saved" in fit()
mode.predict(n=len(predictable_timestamps), series=dummy_series, past/future_covariates=...)
```

Unless your model uses only future_covariates, the forecast will be shorted than the `training_set`. I hope that it correspond to what you wanted.

By curiosity, what do you use this for?

**Comment by ETTAN93** [comment]
Hi,

I was thinking of using the forecasts of the training set for 2 purposes:
i. To identify any possible overfitting in the model
ii. To calculate some sort of correction factor for the forecasts (from test set) by using the mean of the actual values (from train set)/ mean of forecast values (from train set).
 
It does seem a bit odd that the model.predict() method does not allow you to forecast the same timestamp of the training series (or even any timestamp that you want after fitting the model). Is that done for a specific reason?

**Comment by ETTAN93** [comment]
@madtoinou I tried out using the historical forecast method to generate a model forecast for the training set timestamps and found out that different results are returned compared to your method above. Both are using the same train_series, past_cov_series and future_cov_series so I'm not sure why that can be the case.

**Comment by madtoinou** [comment]
> It does seem a bit odd that the model.predict() method does not allow you to forecast the same timestamp of the training series (or even any timestamp that you want after fitting the model). Is that done for a specific reason?

Let's say your model uses `lags=[-1,-2,-3]` or `input_chunk_length=3`, which values is it supposed to uses to predict the first timestamp since there is nothing before? At the moment darts is not allowing user to generate forecast for an arbitrary timestamp to keep a simple API but this feature might come in the future

> I tried out using the historical forecast method to generate a model forecast for the training set timestamps and found out that different results are returned compared to your method above. 

Did you use `retrain=True` (default behavior)? If so, the model is retrained with an expanding window after inferring each `horizon_forecast` (training set increases as the inference get closer to the end of the `series` argument).

To avoid/detect overfitting, I would recommend splitting your data into training, validation and test instead of using the test set directly.
